R/gatherIAMCTimeseries.R

```diff
@@ -0,0 +1,62 @@
+#' gatherIAMCTimeseries
+#'
+#' Pivots periods columns (as they occur in IAMC wide format/timeseries) to the typical `quitte` long list format
+#'
+#' @md
+#' @param df `quitte` data frame in IAMC wide format (i.e. years/periods are columns)
+#' @param quitte_case In IAMC case columns first letters are capitalized, quitte case is lower case. Default is FALSE
+#' @return `quitte` data frame in long IAMC (= `quitte`) format
+#' @importFrom quitte interpolate_missing_periods as.IAMCTimeseries
+#' @importFrom dplyr all_of matches relocate rename_with select
+#' @importFrom tidyr pivot_longer
+#' @importFrom stringr str_to_title str_to_lower
+#' @examples
+#' \dontrun{
+#' df <- quitte_example_data %>% as.IAMCTimeseries()
+#' df %>% gatherIAMCTimeseries()
+#' }
+#' @author Tonn Rüter
+#' @export
+gatherIAMCTimeseries <- function(df, quitte_case = FALSE) {
+  period_transform <- list("period" = as.integer)
+  # IAMC long format default columns. Note: `quitte` uses lower case, IAMC title case
+  default_columns <- c("Model", "Scenario", "Region", "Variable", "Unit", "Period", "Value")
+  # Check if df might already be a proper `quitte` data frame. Suppress warning in case
+  # the vectors have different lengths
+  if (suppressWarnings(all(colnames(df) == tolower(default_columns))))
+    stop("data frame must be in IAMC wide format")
+  # Strip year columns
+  periods <- df %>%
+    select(all_of(matches("^\\d+$"))) %>%
+    colnames()
```

Comment:
This generates a (potentially large) copy of (most of) the data frame.  Just grep the column names.